### DISCUSSION TITLE
Wireguard Mode with a Single Device

### DISCUSSION THREAD

**Discussion by peterstory** [mitmproxy/mitmproxy#5795] [General]
My goal is to monitor outbound traffic from my Mac, using mitmproxy in Wireguard mode running on my Mac (for a few different reasons, I can only use a single device). A diagram showing what I'm trying to do:
```
On my Mac:
Browser/curl/etc. -> Wireguard -> mitmproxy -> the internet
```

Since [mitmproxy seems to have trouble running directly in macOS on Apple Silicon](https://github.com/mitmproxy/mitmproxy/issues/4605), I'm opting to run mitmproxy in Docker:
```
docker run --rm -it \
-v ~/.mitmproxy:/home/mitmproxy/.mitmproxy \
-p 51820:51820/udp -p 127.0.0.1:8081:8081 \
mitmproxy/mitmproxy mitmweb --web-host 0.0.0.0 --mode wireguard
```

I can connect to the web interface without difficultly, but I'm have trouble routing my machine's traffic through mitmproxy. Here is the Wireguard config I'm using:
```
[Interface]
PrivateKey = REDACTED
Address = 10.0.0.1/32
DNS = 10.0.0.53

[Peer]
PublicKey = REDACTED
AllowedIPs = 0.0.0.0/0
Endpoint = 127.0.0.1:51820
```

Running with this config, I see these errors from mitmproxy:
```
[21:51:26.454][10.0.0.1:63222] Closing connection due to inactivity: Client(10.0.0.1:63222, state=open)
[21:51:26.474] Traceback (most recent call last):

  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 395, in wg_handle_udp_datagram
    self.handle_udp_datagram(

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 196, in handle_udp_datagram
    reader = cast(udp.DatagramReader, handler.transports[handler.client].reader)
                                      ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^

KeyError: Client({'id': '…12cb8f', 'peername': ('10.0.0.1', 64050), 'sockname': ('0.0.0.0', 51820), 'timestamp_start': 1670449856.4076679, 'state': <ConnectionState.CLOSED: 0>, 'transport_protocol': 'udp', 'proxy_mode': ProxyMode.parse('wireguard')})

Please lodge a bug report at:
	https://github.com/mitmproxy/mitmproxy/issues
[21:51:26.474] Traceback (most recent call last):

  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 395, in wg_handle_udp_datagram
    self.handle_udp_datagram(

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 196, in handle_udp_datagram
    reader = cast(udp.DatagramReader, handler.transports[handler.client].reader)
                                      ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^

KeyError: Client({'id': '…19ac4b', 'peername': ('10.0.0.1', 61646), 'sockname': ('0.0.0.0', 51820), 'timestamp_start': 1670449856.4279208, 'state': <ConnectionState.CLOSED: 0>, 'transport_protocol': 'udp', 'proxy_mode': ProxyMode.parse('wireguard')})

Please lodge a bug report at:
	https://github.com/mitmproxy/mitmproxy/issues
[21:51:26.475] Traceback (most recent call last):

  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 395, in wg_handle_udp_datagram
    self.handle_udp_datagram(

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 196, in handle_udp_datagram
    reader = cast(udp.DatagramReader, handler.transports[handler.client].reader)
                                      ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^

KeyError: Client({'id': '…23e7b2', 'peername': ('10.0.0.1', 60837), 'sockname': ('0.0.0.0', 51820), 'timestamp_start': 1670449856.4311888, 'state': <ConnectionState.CLOSED: 0>, 'transport_protocol': 'udp', 'proxy_mode': ProxyMode.parse('wireguard')})

Please lodge a bug report at:
	https://github.com/mitmproxy/mitmproxy/issues
[21:51:26.475] Traceback (most recent call last):

  File "/usr/local/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 395, in wg_handle_udp_datagram
    self.handle_udp_datagram(

  File "/usr/local/lib/python3.11/site-packages/mitmproxy/proxy/mode_servers.py", line 196, in handle_udp_datagram
    reader = cast(udp.DatagramReader, handler.transports[handler.client].reader)
                                      ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^

KeyError: Client({'id': '…e8ec7e', 'peername': ('10.0.0.1', 63222), 'sockname': ('0.0.0.0', 51820), 'timestamp_start': 1670449856.436661, 'state': <ConnectionState.CLOSED: 0>, 'transport_protocol': 'udp', 'proxy_mode': ProxyMode.parse('wireguard')})

Please lodge a bug report at:
	https://github.com/mitmproxy/mitmproxy/issues
[21:51:26.483][10.0.0.1:58371] client connect
[21:51:26.485][10.0.0.1:64616] Closing connection due to inactivity: Client(10.0.0.1:64616, state=open)
```

My intuition is that I need to avoid an infinite loop of traffic from mitmproxy going back into Wireguard, but I'm unsure how to do this. 

I have successfully run a similar setup with mitmproxy running in Docker with Wireguard mode on a separate Mac, but it's essential for my use-case that I get everything running on a single device.

**Comment by mhils** [comment]
WireGuard mode really only works well with external devices at the moment. This is unlikely to change unless WireGuard itself develops functionality to exempt individual applications. However, I'm working on an OS proxy mode for Windows that pretty much does what you're looking for. Once we find that this works, we can port the same concept to Linux/macOS using tun interfaces. :-)

**Reply by peterstory** [reply]
Having that functionality would be great! I'm hoping a one-device option will make it possible for me to run a university lab activity using mitmproxy. My intuition is that a one-device solution using Docker and WireGuard will require fewer platform-specific configuration steps than [the old transparent proxying mode](https://docs.mitmproxy.org/stable/howto-transparent/), and will be less likely to be [broken by OS updates](https://github.com/mitmproxy/mitmproxy/issues/4835#issuecomment-952680705).

**Reply by mhils** [reply]
The goal here is to neither require Docker nor WireGuard, but just be able to do `mitmproxy --mode osproxy:curl` to intercept all requests made by binaries named curl. And of course there should also be a generic mode that captures _all_ traffic. :-)

**Reply by peterstory** [reply]
Would osproxy mode be able to run within Docker on macOS and Windows hosts, while capturing traffic from the host?

I've found that running things within Docker is the only surefire way to ensure that all students can get things running in a timely fashion. For example, it seems that [running mitmproxy directly on Apple silicon is challenging](https://github.com/mitmproxy/mitmproxy/issues/4605) – Docker is a good way to avoid surprises like that.

**Reply by mhils** [reply]
For Windows the answer is a definite no, for macOS this may work. Depends on how we eventually end up implementing things.

**Reply by peterstory** [reply]
Looks like [WireSock](https://www.wiresock.net) on Windows supports "selective application tunneling," so using WireSock with mitmproxy running in Docker sounds promising. Unfortunately, it sounds like ["split tunneling" is difficult in recent macOS versions](https://machow2.com/split-tunneling-mac/), so I think I'm unlikely to find a similar solution on macOS.

**Reply by mhils** [reply]
WireSock looks great (and may just work with the existing WireGuard mode - I have not tested it), but our OS proxy mode will not be built on top of it. We'll use WinDivert on Windows, and tun interfaces on Linux and macOS.

**Reply by natalie-o-perret** [reply]
@mhils out of curiosity, is there any news for local redirect support in linux?
> Redirection on Linux is still in its infancy, but [@decathorpe](https://github.com/decathorpe) has plans for an EBPF-based solution.

- https://mitmproxy.org/posts/local-redirect/windows/
- https://mitmproxy.org/posts/local-redirect/macos/

**Reply by mhils** [reply]
Yes, a first version will likely come before EOY. :) https://github.com/mitmproxy/mitmproxy/pull/7278 has the first bits.

**Reply by natalie-o-perret** [reply]
That's awesome!
Can't wait :eyes:

**Reply by spiritualhacker-dev** [reply]
Hi! Appreciate your work on this! Any update on running local on Mac or Linux?

And/or if I am building a test lab that are compatible with mitmproxy, would you recommend Docker, Kubernetes, or something else?
(I'm a bit out of touch with modern test lab configs).

**Reply by mhils** [reply]
Local mode is now available on Windows, Linux, and macOS; see https://mitmproxy.org/tags/local-capture/.
